### DISCUSSION TITLE
using complex/hierarchical state in recoil

### DISCUSSION THREAD

**Discussion by krisrajzlive** [facebookexperimental/Recoil#1164] [Q&A]
I have been using useState and useContext which works fine with the below data structure:

```
const initialState = {
	projectData:
	{
		projectName: '',
		projectCode: '',
		tasks: [],
		resources: []
	}
}

How can I do the same with recoil? Any help would be appreciated much.

**Comment by drarmstr** [comment]
You can use complex objects for Recoil state as well.  It can be convenient to use the updater form of a setter when you just want to update part of the state.  For example:

```jsx
const setProjectState = useSetRecoilState(projectState);

const handleClick = () => {
  setProjectState(oldProjectState => {
    return {
      ...oldProjectState,
      projectData: {
        ...oldProjectState.projectData,
        projectName: 'foo',
      },
    };
  });
};
```

In general, it is often easier to write code dealing with simpler atoms that contain primitives.  It is also easier to manage state persistence or synchronization when dealing with simpler atoms, particularly if you need to worry about backward compatibility if your state hierarchy changes.  Another consideration is performance.  downstream selectors that depend on the atom and React components that subscribe to the atom or derived selectors subscribe based on the atom granularity.  So, changing part of the complex object in an atom may cause other components to re-render even if they only look at other parts of the atom.  Note that some optimizations are in the works that may suppress some of these re-renders.  But, this is something to consider.

**Comment by drarmstr** [comment]
See also #1127
